alibi/explainers/tests/test_anchor_text.py

```diff
@@ -118,3 +163,298 @@ def test_neighbors():
+@pytest.mark.parametrize('punctuation', ['', string.punctuation])
+@pytest.mark.parametrize('stopwords', [[], ['a', 'the', 'but', 'this', 'there', 'those', 'an']])
+@pytest.mark.parametrize('filling_method', ['parallel'])
+@pytest.mark.parametrize('sample_proba', [0.5, 0.6, 0.7])
+def test_lm_stopwords_punctuation(lang_model, punctuation, stopwords, filling_method,
+                                  movie_sentiment_data, sample_proba, nlp):
+    """
+    Checks if the sampling procedure affect the stopwords and the
+    punctuation when is not supposed to.
+    """
+    # unpack test data
+    X_test = movie_sentiment_data['X_test']
+
+    # select 10 samples
+    n = 10
+    np.random.seed(0)
+    idx = np.random.choice(len(X_test), size=n, replace=False)
+
+    # update test dat
+    X_test = [X_test[i] for i in idx]
+    assert len(X_test) == n
+
+    # define perturb opts
+    perturb_opts = {
+        "sample_proba": sample_proba,
+        "punctuation": punctuation,
+        "stopwords": stopwords,
+        "frac_mask_templates": 0.1,
+    }
+
+    # initialize sampler
+    sampler = LanguageModelSampler(model=lang_model, perturb_opts=perturb_opts)
+
+    for i in range(n):
+        text = X_test[i]
+
+        # process test
+        processed = nlp(text)
+        words = [w.text.lower() for w in processed]
+        words = {w: words.count(w) for w in words}
+
+        # set sampler perturb opts
+        sampler.set_text(text)
+
+        # get masks samples
+        raw, data = sampler.create_mask((), num_samples=10, filling_method=filling_method, **perturb_opts)
+
+        for j in range(len(raw)):
+            mask_counts = str(raw[j]).count(lang_model.mask)
+            raw[j] = str(raw[j]).replace(lang_model.mask, '', mask_counts)
+
+            preprocessed = nlp(str(raw[j]))
+            words_masks = [w.text.lower() for w in preprocessed]
+            words_masks = {w: words_masks.count(w) for w in words_masks}
+
+            # check if the stopwords were perturb and they were not supposed to
+            for sw in stopwords:
+                if sw in words:
+                    assert words[sw] == words_masks[sw]
+
+            # check if the punctuation was petrub and it was not supposed to
+            for p in punctuation:
+                if (p in text) and (p != '\''):  # ' is a tricky one as in words don't for which don' is a token
+                    assert text.count(p) == str(raw[j]).count(p)
+
+
+@pytest.mark.parametrize('lang_model', ['DistilbertBaseUncased', 'BertBaseUncased', 'RobertaBase'], indirect=True)
+@pytest.mark.parametrize('head_gt, num_tokens',
+                         [('word1', 2),
+                          ('word1 word2', 3),
+                          ('word1 word2 word3', 4),
+                          ('word1 word2 word3 word4', 5),
+                          ('word1 word2 word3 word4 word5', 6)])
```

Comment:
the `num_token` represents the number of unique tokens in the head. Note that `tail = 'Pneumonoultramicroscopicsilicovolcanoconiosis ' * 100` and the text becomes `text = head_gt + ' ' + tail`. After that, the `head_tail_split` is applied and the number of unique words in the `head` is counted. If the `head_tail_split` functionality works correctly, then it should not include subwords. For example the `head` will be `word1 Pneumonoultramicroscopicsilicovolcanoconiosis ... Pneumonoultramicroscopicsilicovolcanoconiosis` and not `word1 Pneumonoultramicroscopicsilicovolcanoconiosis .... Pneumonoultramicroscopicsilicovolcanoconiosis Pneumonoultramic`. Thus, the number of unique words in the `head` should be 2 (1 from `word1` and 1 from `Pneumonoultramicroscopicsilicovolcanoconiosis`). Otherwise, if the head will also include a subword, the number of unique tokens would be 3 (`word1`, `Pneumonoultramicroscopicsilicovolcanoconiosis` and `Pneumonoultramic`)